executor/new_analyze.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/distsql"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+var _ Executor = &AnalyzeIndexExec{}
+
+func (e *AnalyzeExec) analyzeIndexPushdown(task *analyzeTask) statistics.AnalyzeResult {
+	hist, err := task.src.(*AnalyzeIndexExec).buildHistogram()
+	if err != nil {
+		return statistics.AnalyzeResult{Err: err}
+	}
+	result := statistics.AnalyzeResult{
+		TableID: task.tableInfo.ID,
+		Hist:    []*statistics.Histogram{hist},
+		IsIndex: 1,
+	}
+	if len(hist.Buckets) > 0 {
+		result.Count = hist.Buckets[len(hist.Buckets)-1].Count
+	}
+	return result
+}
+
+// AnalyzeIndexExec represents analyze index push down executor.
+type AnalyzeIndexExec struct {
+	ctx         context.Context
+	tblInfo     *model.TableInfo
+	idxInfo     *model.IndexInfo
+	ranges      []*types.IndexRange
+	concurrency int
+	priority    int
+	analyzePB   *tipb.AnalyzeReq
+	result      distsql.SelectResult
+}
+
+// Schema implements the Executor Schema interface.
+func (e *AnalyzeIndexExec) Schema() *expression.Schema {
+	return expression.NewSchema()
+}
+
+// Open implements the Executor Open interface.
+func (e *AnalyzeIndexExec) Open() error {
```

Comment:
Since `Next` is not called, the `AnalyzeIndexExec` is not used as an ordinary `Executor`.
How about doesn't implement `Open` and `Close` as well, move the code to `buildHistogram`?